In [18]:
import numpy as np
import pandas as pd 
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [19]:
import os

for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/kaggle/input/weather-dataset-rattle-package/weatherAUS.csv


In [20]:
df = pd.read_csv("/kaggle/input/weather-dataset-rattle-package/weatherAUS.csv")
df.head()

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1  2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2  2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4  2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  Humidity3pm  \
0           W           44.0          W  ...        71.0         22.0   
1         WNW           44.0        NNW  ...        44.0         25.0   
2         WSW           46.0          W  ...        38.0         30.0   
3          NE           24.0         SE  ...        45.0         16.0   
4           W           41.0        ENE  ...        82.0         33.0   

   Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  \
0       1007.7       1007.1       8.0       NaN     16.9     21.8         No   
1       1010.6       1007.8       NaN       NaN     17.2     24.3         No   
2       1007.6       1008.7       NaN       2.0     21.0     23.2         No   
3       1017.6       1012.8       NaN       NaN     18.1     26.5         No   
4       1010.8       1006.0       7.0       8.0     17.8     29.7         No   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  

[5 rows x 23 columns]

In [21]:

df.info

<bound method DataFrame.info of               Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
0       2008-12-01   Albury     13.4     22.9       0.6          NaN   
1       2008-12-02   Albury      7.4     25.1       0.0          NaN   
2       2008-12-03   Albury     12.9     25.7       0.0          NaN   
3       2008-12-04   Albury      9.2     28.0       0.0          NaN   
4       2008-12-05   Albury     17.5     32.3       1.0          NaN   
...            ...      ...      ...      ...       ...          ...   
145455  2017-06-21    Uluru      2.8     23.4       0.0          NaN   
145456  2017-06-22    Uluru      3.6     25.3       0.0          NaN   
145457  2017-06-23    Uluru      5.4     26.9       0.0          NaN   
145458  2017-06-24    Uluru      7.8     27.0       0.0          NaN   
145459  2017-06-25    Uluru     14.9      NaN       0.0          NaN   

        Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Humidity9am  \
0            NaN           W

In [22]:
df.isnull().sum()

Date                 0
Location             0
MinTemp           1485
MaxTemp           1261
Rainfall          3261
Evaporation      62790
Sunshine         69835
WindGustDir      10326
WindGustSpeed    10263
WindDir9am       10566
WindDir3pm        4228
WindSpeed9am      1767
WindSpeed3pm      3062
Humidity9am       2654
Humidity3pm       4507
Pressure9am      15065
Pressure3pm      15028
Cloud9am         55888
Cloud3pm         59358
Temp9am           1767
Temp3pm           3609
RainToday         3261
RainTomorrow      3267
dtype: int64

In [23]:
df.duplicated().sum()

0

In [24]:
df.columns

Index(['Date', 'Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation',
       'Sunshine', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm',
       'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm',
       'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am',
       'Temp3pm', 'RainToday', 'RainTomorrow'],
      dtype='object')

In [25]:
df.shape

(145460, 23)

In [26]:
df.dtypes

Date              object
Location          object
MinTemp          float64
MaxTemp          float64
Rainfall         float64
Evaporation      float64
Sunshine         float64
WindGustDir       object
WindGustSpeed    float64
WindDir9am        object
WindDir3pm        object
WindSpeed9am     float64
WindSpeed3pm     float64
Humidity9am      float64
Humidity3pm      float64
Pressure9am      float64
Pressure3pm      float64
Cloud9am         float64
Cloud3pm         float64
Temp9am          float64
Temp3pm          float64
RainToday         object
RainTomorrow      object
dtype: object

In [27]:
df.dropna(subset=['RainToday', 'RainTomorrow'], inplace=True)

In [28]:
train_sample, test_df = train_test_split(df, test_size=0.3, random_state=42)
train_df, val_df = train_test_split(df, test_size=0.25, random_state=42)

print('train_df.shape :', train_df.shape)
print('val_df.shape :', val_df.shape)
print('test_df.shape :', test_df.shape)

train_df.shape : (105590, 23)
val_df.shape : (35197, 23)
test_df.shape : (42237, 23)


In [29]:
year = pd.to_datetime(df.Date).dt.year

train_df = df[year < 2014]
val_df = df[year == 2014]
test_df = df[year > 2014]

print('train_df.shape :', train_df.shape)
print('val_df.shape :', val_df.shape)
print('test_df.shape :', test_df.shape)

train_df.shape : (80807, 23)
val_df.shape : (17181, 23)
test_df.shape : (42799, 23)


In [30]:
train_df.columns

Index(['Date', 'Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation',
       'Sunshine', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm',
       'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm',
       'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am',
       'Temp3pm', 'RainToday', 'RainTomorrow'],
      dtype='object')

In [33]:
input_cols = list(train_df.columns)[1:-1]
target_col = 'RainTomorrow'
print(inp_columns)

['Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation', 'Sunshine', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm', 'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm', 'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am', 'Temp3pm', 'RainToday']


In [34]:
train_inputs = train_df[input_cols].copy()
train_targets = train_df[target_col].copy()

val_inputs = val_df[input_cols].copy()
val_targets = val_df[target_col].copy()

test_inputs = test_df[input_cols].copy()
test_targets = test_df[target_col].copy()

# **Визначаємо категоріальні та числові змінні**

In [35]:
numerical = train_inputs.select_dtypes('number').columns.tolist()
categorical = train_inputs.select_dtypes('object').columns.tolist()

In [36]:
train_inputs[numerical].describe().round(2)

MinTemp   MaxTemp  Rainfall  Evaporation  Sunshine  WindGustSpeed  \
count  80528.00  80644.00  80807.00     52207.00  49444.00       75039.00   
mean      11.92     22.86      2.46         5.17      7.57          40.27   
std        6.34      6.93      8.81         3.78      3.78          13.73   
min       -8.50     -3.80      0.00         0.00      0.00           6.00   
25%        7.40     17.70      0.00         2.40      4.80          31.00   
50%       11.70     22.20      0.00         4.40      8.40          39.00   
75%       16.50     27.70      0.80         7.00     10.60          48.00   
max       33.90     48.10    371.00        82.40     14.30         135.00   

       WindSpeed9am  WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  \
count      80200.00      80016.00     79840.00     79815.00     73628.00   
mean          14.15         18.83        68.81        51.72      1017.43   
std            9.01          8.90        18.86        20.65         7.08   
min            0.00          0.00         0.00         0.00       980.50   
25%            7.00         13.00        57.00        37.00      1012.70   
50%           13.00         19.00        70.00        52.00      1017.40   
75%           20.00         24.00        83.00        66.00      1022.30   
max           87.00         87.00       100.00       100.00      1041.00   

       Pressure3pm  Cloud9am  Cloud3pm   Temp9am   Temp3pm  
count     73608.00  52542.00  51588.00  80281.00  80261.00  
mean       1015.06      4.29      4.43     16.73     21.39  
std           7.00      2.86      2.68      6.39      6.78  
min         979.00      0.00      0.00     -5.90     -5.10  
25%        1010.30      1.00      2.00     12.10     16.40  
50%        1015.10      5.00      5.00     16.40     20.70  
75%        1019.90      7.00      7.00     21.20     26.00  
max        1039.60      9.00      9.00     38.90     46.10

In [37]:
train_inputs[categorical].nunique()

Location       49
WindGustDir    16
WindDir9am     16
WindDir3pm     16
RainToday       2
dtype: int64

# Заповнюємо відсутні числові дані

In [39]:
imputer = SimpleImputer(strategy = 'mean')
imputer.fit(train_inputs[numerical])

train_inputs[numerical] = imputer.transform(train_inputs[numerical])
val_inputs[numerical] = imputer.transform(val_inputs[numerical])
test_inputs[numerical] = imputer.transform(test_inputs[numerical])

In [40]:
train_inputs[numerical].isna().sum()

MinTemp          0
MaxTemp          0
Rainfall         0
Evaporation      0
Sunshine         0
WindGustSpeed    0
WindSpeed9am     0
WindSpeed3pm     0
Humidity9am      0
Humidity3pm      0
Pressure9am      0
Pressure3pm      0
Cloud9am         0
Cloud3pm         0
Temp9am          0
Temp3pm          0
dtype: int64

# Маштабуємо

In [42]:
scaler = MinMaxScaler()
scaler.fit(train_df[numerical])

train_inputs[numerical] = scaler.transform(train_inputs[numerical])
val_inputs[numerical] = scaler.transform(val_inputs[numerical])
test_inputs[numerical] = scaler.transform(test_inputs[numerical])

In [43]:
train_inputs[numerical].describe()

MinTemp       MaxTemp      Rainfall   Evaporation      Sunshine  \
count  80807.000000  80807.000000  80807.000000  80807.000000  80807.000000   
mean       0.481721      0.513656      0.006635      0.062773      0.529558   
std        0.149295      0.133398      0.023758      0.036881      0.206833   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.375000      0.414258      0.000000      0.043689      0.510490   
50%        0.476415      0.500963      0.000000      0.062773      0.529558   
75%        0.589623      0.606936      0.002156      0.067961      0.650350   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

       WindGustSpeed  WindSpeed9am  WindSpeed3pm   Humidity9am   Humidity3pm  \
count   80807.000000  80807.000000  80807.000000  80807.000000  80807.000000   
mean        0.265633      0.162676      0.216387      0.688060      0.517183   
std         0.102587      0.103183      0.101765      0.187472      0.205191   
min         0.000000      0.000000      0.000000      0.000000      0.000000   
25%         0.193798      0.080460      0.149425      0.570000      0.370000   
50%         0.255814      0.149425      0.218391      0.700000      0.520000   
75%         0.310078      0.218391      0.275862      0.830000      0.660000   
max         1.000000      1.000000      1.000000      1.000000      1.000000   

        Pressure9am   Pressure3pm      Cloud9am      Cloud3pm       Temp9am  \
count  80807.000000  80807.000000  80807.000000  80807.000000  80807.000000   
mean       0.610391      0.595115      0.476711      0.491904      0.505149   
std        0.111777      0.110184      0.256300      0.238043      0.142151   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.542149      0.524752      0.333333      0.333333      0.401786   
50%        0.610391      0.595115      0.476711      0.491904      0.500000   
75%        0.680992      0.665017      0.666667      0.666667      0.604911   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

            Temp3pm  
count  80807.000000  
mean       0.517320  
std        0.131958  
min        0.000000  
25%        0.419922  
50%        0.505859  
75%        0.607422  
max        1.000000

In [44]:
encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
encoder.fit(train_inputs[categorical])
encoder.categories_

[array(['Adelaide', 'Albany', 'Albury', 'AliceSprings', 'BadgerysCreek',
        'Ballarat', 'Bendigo', 'Brisbane', 'Cairns', 'Canberra', 'Cobar',
        'CoffsHarbour', 'Dartmoor', 'Darwin', 'GoldCoast', 'Hobart',
        'Katherine', 'Launceston', 'Melbourne', 'MelbourneAirport',
        'Mildura', 'Moree', 'MountGambier', 'MountGinini', 'Newcastle',
        'Nhil', 'NorahHead', 'NorfolkIsland', 'Nuriootpa', 'PearceRAAF',
        'Penrith', 'Perth', 'PerthAirport', 'Portland', 'Richmond', 'Sale',
        'SalmonGums', 'Sydney', 'SydneyAirport', 'Townsville',
        'Tuggeranong', 'Uluru', 'WaggaWagga', 'Walpole', 'Watsonia',
        'Williamtown', 'Witchcliffe', 'Wollongong', 'Woomera'],
       dtype=object),
 array(['E', 'ENE', 'ESE', 'N', 'NE', 'NNE', 'NNW', 'NW', 'S', 'SE', 'SSE',
        'SSW', 'SW', 'W', 'WNW', 'WSW', nan], dtype=object),
 array(['E', 'ENE', 'ESE', 'N', 'NE', 'NNE', 'NNW', 'NW', 'S', 'SE', 'SSE',
        'SSW', 'SW', 'W', 'WNW', 'WSW', nan], dtype=object),
 ar

In [45]:

encoded_cols = list(encoder.get_feature_names_out(categorical))
print(encoded_cols)

['Location_Adelaide', 'Location_Albany', 'Location_Albury', 'Location_AliceSprings', 'Location_BadgerysCreek', 'Location_Ballarat', 'Location_Bendigo', 'Location_Brisbane', 'Location_Cairns', 'Location_Canberra', 'Location_Cobar', 'Location_CoffsHarbour', 'Location_Dartmoor', 'Location_Darwin', 'Location_GoldCoast', 'Location_Hobart', 'Location_Katherine', 'Location_Launceston', 'Location_Melbourne', 'Location_MelbourneAirport', 'Location_Mildura', 'Location_Moree', 'Location_MountGambier', 'Location_MountGinini', 'Location_Newcastle', 'Location_Nhil', 'Location_NorahHead', 'Location_NorfolkIsland', 'Location_Nuriootpa', 'Location_PearceRAAF', 'Location_Penrith', 'Location_Perth', 'Location_PerthAirport', 'Location_Portland', 'Location_Richmond', 'Location_Sale', 'Location_SalmonGums', 'Location_Sydney', 'Location_SydneyAirport', 'Location_Townsville', 'Location_Tuggeranong', 'Location_Uluru', 'Location_WaggaWagga', 'Location_Walpole', 'Location_Watsonia', 'Location_Williamtown', 'Loca

In [47]:
train_inputs[encoded_cols] = encoder.transform(train_inputs[categorical])
val_inputs[encoded_cols] = encoder.transform(val_inputs[categorical])
test_inputs[encoded_cols] = encoder.transform(test_inputs[categorical])

In [48]:
pd.set_option('display.max_columns', None)

In [49]:
test_inputs

Location   MinTemp   MaxTemp  Rainfall  Evaporation  Sunshine  \
2133     Albury  0.469340  0.718690       0.0     0.062773  0.529558   
2134     Albury  0.566038  0.836224       0.0     0.062773  0.529558   
2135     Albury  0.603774  0.811175       0.0     0.062773  0.529558   
2136     Albury  0.813679  0.710983       0.0     0.062773  0.529558   
2137     Albury  0.648585  0.751445       0.0     0.062773  0.529558   
...         ...       ...       ...       ...          ...       ...   
145454    Uluru  0.283019  0.493256       0.0     0.062773  0.529558   
145455    Uluru  0.266509  0.524085       0.0     0.062773  0.529558   
145456    Uluru  0.285377  0.560694       0.0     0.062773  0.529558   
145457    Uluru  0.327830  0.591522       0.0     0.062773  0.529558   
145458    Uluru  0.384434  0.593449       0.0     0.062773  0.529558   

       WindGustDir  WindGustSpeed WindDir9am WindDir3pm  WindSpeed9am  \
2133           WSW       0.186047        ESE          W      0.080460   
2134            NE       0.387597        ESE        ESE      0.103448   
2135           NNE       0.325581         NE          N      0.229885   
2136           NNE       0.271318        ESE          W      0.080460   
2137             E       0.209302        SSE         SE      0.080460   
...            ...            ...        ...        ...           ...   
145454           E       0.193798        ESE          E      0.172414   
145455           E       0.193798         SE        ENE      0.149425   
145456         NNW       0.124031         SE          N      0.149425   
145457           N       0.240310         SE        WNW      0.103448   
145458          SE       0.170543        SSE          N      0.149425   

        WindSpeed3pm  Humidity9am  Humidity3pm  Pressure9am  Pressure3pm  \
2133        0.126437         0.45         0.14     0.545455     0.528053   
2134        0.103448         0.45         0.12     0.586777     0.551155   
2135        0.229885         0.35         0.19     0.618182     0.549505   
2136        0.080460         0.46         0.37     0.547107     0.546205   
2137        0.103448         0.60         0.34     0.609917     0.589109   
...              ...          ...          ...          ...          ...   
145454      0.149425         0.59         0.27     0.730579     0.696370   
145455      0.126437         0.51         0.24     0.728926     0.681518   
145456      0.103448         0.56         0.21     0.710744     0.661716   
145457      0.103448         0.53         0.24     0.669421     0.623762   
145458      0.080460         0.51         0.24     0.642975     0.618812   

        Cloud9am  Cloud3pm   Temp9am   Temp3pm RainToday  Location_Adelaide  \
2133    0.476711  0.491904  0.600446  0.738281        No                0.0   
2134    0.476711  0.491904  0.703125  0.845703        No                0.0   
2135    0.476711  0.491904  0.783482  0.822266        No                0.0   
2136    0.888889  0.555556  0.743304  0.703125        No                0.0   
2137    0.888889  0.491904  0.703125  0.734375        No                0.0   
...          ...       ...       ...       ...       ...                ...   
145454  0.476711  0.491904  0.341518  0.507812        No                0.0   
145455  0.476711  0.491904  0.357143  0.537109        No                0.0   
145456  0.476711  0.491904  0.375000  0.578125        No                0.0   
145457  0.476711  0.491904  0.410714  0.609375        No                0.0   
145458  0.333333  0.222222  0.468750  0.607422        No                0.0   

        Location_Albany  Location_Albury  Location_AliceSprings  \
2133                0.0              1.0                    0.0   
2134                0.0              1.0                    0.0   
2135                0.0              1.0                    0.0   
2136                0.0              1.0                    0.0   
2137                0.0              1.0                    0.0   
.

# Створюємо логістичну регресію

In [51]:
model = LogisticRegression(solver='liblinear')
model.fit(train_inputs[numerical + encoded_cols], train_targets)

LogisticRegression(solver='liblinear')

# Оцінюємо модель

In [52]:
X_train = train_inputs[numerical + encoded_cols]
X_val = val_inputs[numerical + encoded_cols]
X_test = test_inputs[numerical + encoded_cols]

In [53]:
train_preds = model.predict(X_train)
train_preds

array(['No', 'No', 'No', ..., 'No', 'No', 'No'], dtype=object)

In [54]:

pd.Series(train_preds).value_counts(normalize=True)

No     0.836945
Yes    0.163055
Name: proportion, dtype: float64

In [55]:
train_probs = model.predict_proba(X_train)
train_probs

array([[0.94358421, 0.05641579],
       [0.93998488, 0.06001512],
       [0.96164668, 0.03835332],
       ...,
       [0.98273706, 0.01726294],
       [0.87291136, 0.12708864],
       [0.92823898, 0.07176102]])

In [58]:
accuracy_score(train_targets, train_preds)

0.8511886346479884

In [59]:
errors = pd.DataFrame({'target': train_targets, 'prediction': train_preds})
errors

target prediction
0          No         No
1          No         No
2          No         No
3          No         No
4          No         No
...       ...        ...
144183     No         No
144184     No         No
144185     No         No
144186     No         No
144187     No         No

[80807 rows x 2 columns]

In [60]:
confusion_matrix(train_targets, train_preds)

array([[59031,  3425],
       [ 8600,  9751]])

# Будуємо classification_report

In [61]:
test_preds = model.predict(X_test)

print(classification_report(test_targets, test_preds))

              precision    recall  f1-score   support

          No       0.87      0.95      0.91     33396
         Yes       0.73      0.48      0.58      9403

    accuracy                           0.85     42799
   macro avg       0.80      0.71      0.74     42799
weighted avg       0.84      0.85      0.83     42799



**Висновок

Точність логістичної регресії - 85%

A класс No передбачається краще ніж клас Yes**